tinygrad/tensor.py

```diff
@@ -24,36 +24,24 @@ def __enter__(self):
     if DEBUG: self.st = time.time()
   def __exit__(self, *junk):
     if DEBUG:
-      if cl_queue is not None:
-        cl_queue.finish()
       et = (time.time()-self.st)*1000.
       debug_counts[self.name] += 1
       debug_times[self.name] += et
       print(f"{self.name:>20} : {et:>7.2f} ms {[y.shape for y in self.x]}")
 
 # **** GPU functions ****
 
-cl_ctx, cl_queue = None, None
-def require_init_gpu():
-  if not GPU: raise Exception("No GPU Support, install pyopencl")
-  global cl_ctx, cl_queue
-  if cl_queue is None:
-    devices = cl.get_platforms()[0].get_devices(device_type=cl.device_type.GPU)
-    if len(devices) == 0:
-      devices = cl.get_platforms()[0].get_devices(device_type=cl.device_type.CPU)
-    cl_ctx = cl.Context(devices=devices)
-    # this is an in-order command queue
-    cl_queue = cl.CommandQueue(cl_ctx)
+api, thr = None, None
+
 
 class GPUBuffer:
   def __init__(self, shape, hostbuf=None):
     self.shape, self.dtype = tuple(shape), np.float32
     self.cl = hostbuf.cl if isinstance(hostbuf, GPUBuffer) else \
```

Comment:
This line throws an error if hostbuf is not contiguous. Would recommend adding something like:
```python
self.cl = hostbuf.cl if isinstance(hostbuf, GPUBuffer) else \
      thr.to_device(np.ascontiguousarray(hostbuf, dtype=np.float32) if hostbuf is not None else np.zeros(shape, np.float32))
```
Haven't checked how expensive `ascontiguousarray()` is though...

Reply:
Could you provide an example with not-contiguous array?

Reply:
For instance with running inference on YOLO, a lot of times there are some preprocessing on the input data like this which causes the input data to not be contiguous:
```python
def infer(model, img):
  img = np.array(img)
  img = imresize(img, 608, 608) # img is contiguous
  # img = resize(img)
  img = img[:,:,::-1].transpose((2,0,1)) # img becomes not contiguous here
  img = img[np.newaxis,:,:,:]/255.0

  prediction = model.forward(Tensor(img))
  return prediction
```

Things like this will probably occur quite a lot, so I think it would be more user friendly if there were checks in the GPUBuffer class making the inputs contiguous, that way the user doesn't have to always make inputs contiguous...

Reply:
Thank you, I'll try to fix it tomorrow!

Reply:
Frankly, I can not reproduce this error
```python
import os
import numpy as np
os.environ['GPU'] = '1'
from tinygrad.tensor import Tensor
arr = np.arange(12).reshape(3,4)
print(arr.flags)
x = Tensor(arr)
arr = arr.T
print(arr.flags)
y = Tensor(arr)
```
This works just fine
Output:
```
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

warning, (3, 4) isn't float32
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
Process finished with exit code 0
```

Reply:
Try this, this produces error for me without the suggested fix.

```python
import os
import numpy as np
os.environ['GPU'] = '1'
from tinygrad.tensor import Tensor

x = np.random.rand(416, 416, 3)
print(x.data.contiguous) # True
x = x[:,:,::-1].transpose((2,0,1))
print(x.data.contiguous) # False
y = Tensor(x)
```

Reply:
I've added fix with ```np.ravel()```
What do you think?

Reply:
And second version like this
Which is better?
``` python
 thr.to_device((hostbuf.astype(np.float32).ravel() if not hostbuf.data.contiguous \
                       else hostbuf.astype(np.float32)) if hostbuf is not None else np.zeros(shape, np.float32))
```

Reply:
Well, I mean, ravel() flattens the input, but it doesn't seem to have any effect so I guess it's the way to go, I also made some benchmarks, it runs faster than `ascontiguousarray()`, so your solution is the best!

Reply:
```.ravel()``` with or withour check is better, what do you think?

Reply:
I think the one with the check is more readable, it's easy to understand why `.ravel()` is called.

Reply:
Thank you for your time!